python/ray/data/_internal/execution/resource_manager.py

```diff
@@ -121,3 +136,129 @@ def get_downstream_fraction(self, op: PhysicalOperator) -> float:
+
+    def __init__(self, resource_manager: ResourceManager):
+        self._resource_manager = resource_manager
+
+    def on_global_limits_updated(self, global_limits: ExecutionResources):
+        """Callback when the global limits are updated."""
+        pass
+
+    @abstractmethod
+    def update_usages(self) -> ExecutionResources:
+        """Callback to update resource usages."""
+        ...
+
+    @abstractmethod
+    def get_op_limits(self, op: PhysicalOperator) -> ExecutionResources:
+        """Get the limit of resources that the given op can use at the current time."""
+        ...
+
+
+class ReservationOpResourceLimiter(OpResourceLimiter):
+    """An OpResourceLimiter implementation that reserves resources for each operator.
+
+    This class reserves memory and CPU resources for map operators, and consider runtime
+    resource usages to limit the resources that each operator can use.
+
+    It works in the following way:
+    1. Currently we only limit map operators. Non-map operators get unlimited resources.
+    2. For each map operator, we reserve `reservation_ratio * global_resources /
+        num_map_ops` resources. The remaining are shared among all map operators.
+    3. In each scheduling iteration, each map operator will get "remaining of their own
+       reserved resources" + "remaining of shared resources / num_map_ops" resources.
+
+    The `reservation_ratio` is set to 50% by default. Users can tune this value to
+    adjust how aggressive or conservative the resource allocation is. A higher value
+    will make the resource allocation more even, but may lead to underutilization and
+    worse performance. And vice versa.
+    """
+
+    def __init__(self, resource_manager: ResourceManager, reservation_ratio: float):
+        super().__init__(resource_manager)
+        self._reservation_ratio = reservation_ratio
+        assert 0.0 <= self._reservation_ratio <= 1.0
+        # We only limit map operators.
+        self._eligible_ops = [
+            op for op in self._resource_manager._topology if isinstance(op, MapOperator)
+        ]
+        # Per-op reserved resources.
+        self._op_reserved: Dict[PhysicalOperator, ExecutionResources] = {}
+        # Total shared resources.
+        self._total_shared = ExecutionResources.zero()
+        # Resource limits for each operator.
+        self._op_limits: Dict[PhysicalOperator, ExecutionResources] = {}
+
+    def on_global_limits_updated(self, global_limits: ExecutionResources):
+        self._total_shared = global_limits.scale(1.0 - self._reservation_ratio)
+
+        default_reserved = global_limits.scale(
+            self._reservation_ratio / len(self._eligible_ops)
+        )
+        for op in self._eligible_ops:
+            # Make sure the reserved resources are at least to allow
+            # one task.
+            self._op_reserved[op] = default_reserved.max(
+                op.incremental_resource_usage()
+            )
+
+    def update_usages(self):
```

Comment:
Is `update_usages` always called whenever the global limits change?

Reply:
it doesn't need to. It only needs to be updated when `ReousrceManager.update_usages` is called.